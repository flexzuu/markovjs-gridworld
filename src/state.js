// @flow
import { Map, List } from 'immutable'
import type { State } from './types'

export const createBoard = ({ size }) =>
  List().setSize(size).map(() => List().setSize(size).map(() => null))

export const placeRandomTilesFactory = random => ({ numberOfRandomTiles = 0, board }) => {
  if (numberOfRandomTiles === 0 || board.flatten().filter(tile => !tile).size === 0) {
    return board
  }
  // place one random tile
  // we place a 2 or 4
  const decider = random(0, 1, true)
  const value = decider < 0.9 ? 2 : 4
  const randomPosition = [random(board.size - 1), random(board.size - 1)]
  const canPlaceTile = board.getIn(randomPosition) === null
  // we need to place more tiles
  if (canPlaceTile) {
    const boardWithRandomTile = board.setIn(randomPosition, value)
    return placeRandomTilesFactory(random)({
      numberOfRandomTiles: numberOfRandomTiles - 1,
      board: boardWithRandomTile,
    })
  }
  // we cant place a tile here so we simply try again
  return placeRandomTilesFactory(random)({ numberOfRandomTiles, board })
}

export const init = ({ size, startTiles, random }): State => {
  const emptyBoard = createBoard({ size })
  const placeRandomTiles = placeRandomTilesFactory(random)
  const board = placeRandomTiles({ numberOfRandomTiles: startTiles, board: emptyBoard })
  return Map({
    board,
    score: 0,
    final: false,
  })
}
