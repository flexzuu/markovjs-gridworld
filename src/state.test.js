import { createBoard, placeRandomTilesFactory, init } from './state'
import { random } from 'lodash'
import { List } from 'immutable'

test('create empty board', () => {
  expect(createBoard({ size: 4 })).toMatchSnapshot()
})
describe('place random Tile', () => {
  test('place no tile', () => {
    const board = createBoard({ size: 4 })
    const pseudoRandom = (x, y, bool) => {
      if (x === 0 && y === 1 && bool) return 0.99
      return board.size
    }
    const placeRandomTiles = placeRandomTilesFactory(pseudoRandom)
    expect(placeRandomTiles({ board })).toMatchSnapshot()
  })
  test('place 1 tile with value 4 in bottom right', () => {
    const board = createBoard({ size: 4 })
    const pseudoRandom = (x, y, bool) => {
      if (x === 0 && y === 1 && bool) return 0.99
      return x
    }
    const placeRandomTiles = placeRandomTilesFactory(pseudoRandom)
    expect(placeRandomTiles({ board, numberOfRandomTiles: 1 })).toMatchSnapshot()
  })
  test('place 1 tile with value 2 in bottom right', () => {
    const board = createBoard({ size: 4 })
    const pseudoRandom = (x, y, bool) => {
      if (x === 0 && y === 1 && bool) return 0.1
      return x
    }
    const placeRandomTiles = placeRandomTilesFactory(pseudoRandom)
    expect(placeRandomTiles({ board, numberOfRandomTiles: 1 })).toMatchSnapshot()
  })
  test('place 1 tile with value 2 in top left', () => {
    const board = createBoard({ size: 4 })
    const pseudoRandom = (x, y, bool) => {
      if (x === 0 && y === 1 && bool) return 0.1
      return 0
    }
    const placeRandomTiles = placeRandomTilesFactory(pseudoRandom)
    expect(placeRandomTiles({ board, numberOfRandomTiles: 1 })).toMatchSnapshot()
  })
  test('place 3 tiles', () => {
    const board = createBoard({ size: 4 })
    const placeRandomTiles = placeRandomTilesFactory(random)
    expect(
      placeRandomTiles({ board, numberOfRandomTiles: 3 }).flatten().filter(tile => tile).size,
    ).toBe(3)
  })
  test('dont crash on full board', done => {
    const board = List([[2, 2], [2, 2]])
    const placeRandomTiles = placeRandomTilesFactory(random)
    placeRandomTiles({ board, numberOfRandomTiles: 1 })
    done()
  })
})

describe('init state', () => {
  test('with a board with 2 tiles', () => {
    const state = init({ size: 4, startTiles: 2, random })
    expect(state.get('board').flatten().filter(tile => tile).size).toBe(2)
  })
  test('with score of 0', () => {
    const state = init({ size: 4, startTiles: 2, random })
    expect(state.get('score')).toBe(0)
  })
  test('with a empty board', () => {
    expect(init({ size: 4, startTiles: 0, random })).toMatchSnapshot()
  })
})
