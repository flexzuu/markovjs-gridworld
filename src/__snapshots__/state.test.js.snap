// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create empty board 1`] = `
Immutable.List [
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
]
`;

exports[`init state with a empty board 1`] = `
Immutable.Map {
  "board": Immutable.List [
    Immutable.List [
      null,
      null,
      null,
      null,
    ],
    Immutable.List [
      null,
      null,
      null,
      null,
    ],
    Immutable.List [
      null,
      null,
      null,
      null,
    ],
    Immutable.List [
      null,
      null,
      null,
      null,
    ],
  ],
  "score": 0,
  "final": false,
}
`;

exports[`place random Tile place 1 tile with value 2 in bottom right 1`] = `
Immutable.List [
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    2,
  ],
]
`;

exports[`place random Tile place 1 tile with value 2 in top left 1`] = `
Immutable.List [
  Immutable.List [
    2,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
]
`;

exports[`place random Tile place 1 tile with value 4 in bottom right 1`] = `
Immutable.List [
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    4,
  ],
]
`;

exports[`place random Tile place no tile 1`] = `
Immutable.List [
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
  Immutable.List [
    null,
    null,
    null,
    null,
  ],
]
`;
